runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -1174,23 +1029,400 @@ static iree_status_t iree_hal_hip_device_queue_write(
+        data->signal_semaphore_list.semaphores[i], &unused_return_value));
+  }
+
+  // 4) Free the iree_hal_hip_device_semaphore_submit_callback_data_t and
+  // the resource set attached.
+  iree_hal_resource_set_free(data->resource_set);
+  iree_slim_mutex_deinitialize(&data->status_mutex);
+  iree_allocator_free(device->host_allocator, data);
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+static iree_status_t iree_hal_hip_device_execute_now(
+    iree_hal_hip_device_semaphore_submit_callback_data_t* data) {
+  IREE_TRACE_ZONE_BEGIN(z0);
+  IREE_ASSERT_EQ(iree_math_count_ones_u64(data->queue_affinity), 1,
+                 "Cannot execute a command buffer on more than one queue");
+  iree_hal_hip_device_t* device = data->device;
+  iree_status_t status = iree_ok_status();
+
+  // If we had a semaphore failure, then we should propagate it
+  // but not run anything
+  if (!iree_status_is_ok(data->status)) {
+    for (iree_host_size_t i = 0; i < data->signal_semaphore_list.count; ++i) {
+      iree_hal_semaphore_fail(data->signal_semaphore_list.semaphores[i],
+                              iree_status_clone(data->status));
+    }
+    return data->status;
+  }
+
+  uint32_t idx = iree_math_count_trailing_zeros_u64(data->queue_affinity);
+  IREE_TRACE_ZONE_APPEND_VALUE_I64(z0, idx);
+
+  IREE_RETURN_AND_END_ZONE_IF_ERROR(
+      z0,
+      IREE_HIP_RESULT_TO_STATUS(
+          data->device->hip_symbols,
+          hipCtxPushCurrent(data->device->topology.devices[idx].hip_context)));
+
+  IREE_TRACE_ZONE_BEGIN_NAMED(z2, "Enqueue semaphore wait");
+  // TODO(awoloszyn): Because of how hip works, if we only have a single
+  // physical device in the hip_device we could avoid waiting on any of these
+  // semaphores, we are guaranteed to have waits, but if we want this
+  // to work across multiple device/streams, we need these waits.
+  for (iree_host_size_t i = 0;
+       i < data->wait_semaphore_list.count && iree_status_is_ok(status); ++i) {
+    iree_hal_hip_event_t* event;
+    IREE_RETURN_AND_END_ZONE_IF_ERROR(
+        z0, iree_hal_hip_semaphore_get_hip_event(
+                data->wait_semaphore_list.semaphores[i],
+                data->wait_semaphore_list.payload_values[i],
+                data->device->topology.devices[idx].device_event_pool, false,
+                &event));
+    // If we don't have an event, then we don't have to wait for it since it
+    // has already been signaled on the host.
+    if (!event) {
+      continue;
+    }
+
+    IREE_TRACE_ZONE_BEGIN_NAMED(z1, "hipStreamWaitEvent");
+    status = iree_status_join(
+        status, IREE_HIP_RESULT_TO_STATUS(
+                    data->device->hip_symbols,
+                    hipStreamWaitEvent(
+                        device->topology.devices[idx].hip_dispatch_stream,
+                        iree_hal_hip_event_handle(event), 0)));
+    iree_hal_hip_event_release(event);
+    IREE_TRACE_ZONE_END(z1);
+  }
+  IREE_TRACE_ZONE_END(z2);
```

Comment:
if you end up with multiple zones in a single function that don't nest properly then you should name them - here you have z0 > z2 > z1. z_enqueue_waits and z_wait_event would make it clear what they are and how they nest